amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManagerTest.java

```diff
@@ -0,0 +1,365 @@
+
+        // Assert that all the leases now has hashRanges set.
+        for(Lease lease : multiStreamLeases) {
+            Assert.assertNotNull(lease.hashKeyRangeForLease());
+        }
+    }
+
+    @Test
+    public void testIfMissingHashRangeInformationIsFilledBeforeEvaluatingForShardSyncInHoleScenario() {
+        ShardSyncTaskManager shardSyncTaskManager = mock(ShardSyncTaskManager.class);
+        ShardDetector shardDetector = mock(ShardDetector.class);
+        when(shardSyncTaskManagerProvider.apply(any())).thenReturn(shardSyncTaskManager);
+        when(shardSyncTaskManager.shardDetector()).thenReturn(shardDetector);
+
+        final int[] shardCounter = { 0 };
+        List<HashKeyRangeForLease> hashKeyRangeForLeases = new ArrayList<HashKeyRangeForLease>() {{
+            add(deserialize(MIN_HASH_KEY.toString(), "1"));
+            add(deserialize("2", "3"));
+            add(deserialize("5", "20")); // Hole between 3 and 5
+            add(deserialize("21", "23"));
+            add(deserialize("24", MAX_HASH_KEY.toString()));
+        }};
+
+        List<Shard> kinesisShards = hashKeyRangeForLeases.stream()
+                .map(hashKeyRangeForLease -> Shard.builder().shardId("shard-" + (++shardCounter[0])).hashKeyRange(
+                        HashKeyRange.builder().startingHashKey(hashKeyRangeForLease.serializedStartingHashKey())
+                                .endingHashKey(hashKeyRangeForLease.serializedEndingHashKey()).build()).build())
+                .collect(Collectors.toList());
+
+        when(shardDetector.listShards()).thenReturn(kinesisShards);
+
+        final int[] leaseCounter = { 0 };
+        List<Lease> multiStreamLeases = hashKeyRangeForLeases.stream().map(hashKeyRangeForLease -> {
+            MultiStreamLease lease = new MultiStreamLease();
+            lease.leaseKey(MultiStreamLease.getLeaseKey(streamIdentifier.serialize(), "shard-"+(++leaseCounter[0])));
+            lease.shardId("shard-"+(leaseCounter[0]));
+            // Setting the hashrange only for last two leases
+            if(leaseCounter[0] >= 3) {
+                lease.hashKeyRange(hashKeyRangeForLease);
+            }
+            lease.checkpoint(ExtendedSequenceNumber.TRIM_HORIZON);
+            return lease;
+        }).collect(Collectors.toList());
+
+        // Assert that shard sync should never trigger
+        IntStream.range(1, CONSECUTIVE_HOLES_FOR_TRIGGERING_RECOVERY).forEach(i -> Assert
+                .assertFalse(periodicShardSyncManager.checkForShardSync(streamIdentifier, multiStreamLeases).shouldDoShardSync()));
+        Assert.assertTrue(periodicShardSyncManager.checkForShardSync(streamIdentifier, multiStreamLeases).shouldDoShardSync());
+
+        // Assert that all the leases now has hashRanges set.
+        for(Lease lease : multiStreamLeases) {
+            Assert.assertNotNull(lease.hashKeyRangeForLease());
+        }
+    }
+
```

Comment:
We need to add test case based on the scenario of splits and merges and validate the completeness of the lease table

Reply:
Added tests for random split, merged and case of pending parent completions.

Reply:
Thanks for adding various test cases looks good.